### DISCUSSION TITLE
Vectorial material property from input

### DISCUSSION THREAD

**Discussion by gabburgio** [idaholab/moose#23612] [Q&A General] [answered]
Hello everybody

I am trying to define a material that provides as a property a vector of diffusivity values for a set of n coupled reaction-diffusion equation, n determined by the input file.
These values should be inserted as a parameter in the material declaration.

The way I have been trying to implement this is:

```

#pragma once

#include "ADMaterial.h"


class othermaterial : public ADMaterial
{
public:
  static InputParameters validParams();

  othermaterial(const InputParameters & parameters);

protected:
  virtual void computeQpProperties() override;

  const std::vector<Real> & _vettore;

  ADMaterialProperty<std::vector<Real>> & _diff_vector;

};
```

for the header file and:
```

#include "othermaterial.h"

registerMooseObject("pertyApp", othermaterial);

InputParameters
othermaterial::validParams()
{
  InputParameters params = ADMaterial::validParams();
  params.addRequiredParam<std::vector<Real>>(
      "vettore", "The vector of diffusivity values");
  return params;
}

othermaterial::othermaterial(const InputParameters & parameters)
  : ADMaterial(parameters),
    _diff_vector(declareADProperty<std::vector<Real>>("diff_vector")),
    _vettore(getParam<std::vector<Real>>("vettore"))

{
}

void
othermaterial::computeQpProperties()
{

  for (int i = 0; i < _vettore.size(); ++i) {
    _diff_vector[_qp][i] = _vettore[i];
  }
}
```

For the source file. However by declaring for example:

```
  [Materials]
    [./example]
      type = othermaterial
      vettore = '1 100'
      block = 'domain'
    [../]
  []  [Materials]
    [./example]
      type = othermaterial
      vettore = '1 100'
      block = 'domain'
    [../]
  []
```

in my input file I get a segfault with the following backtrace:

`Thread 1 "perty-dbg" received signal SIGSEGV, Segmentation fault.
0x00007ffff7f3e1cf in MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true>::operator=<double> (this=0x0, scalar=@0x5555559bf050: 1) at /home/gburgio/mambaforge3/envs/moose/libmesh/include/metaphysicl/dualnumber.h:195
195	  _val = scalar;
(gdb) bt
#0  0x00007ffff7f3e1cf in MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true>::operator=<double> (this=0x0, scalar=@0x5555559bf050: 1) at /home/gburgio/mambaforge3/envs/moose/libmesh/include/metaphysicl/dualnumber.h:195
#1  0x00007ffff7f4fb65 in othermaterial::computeQpProperties (this=0x5555559dac80) at /home/gburgio/projects/perty/src/materials/othermaterial.C:27
#2  0x00007ffff64f4034 in Material::computeProperties (this=0x5555559dac80) at /home/gburgio/projects/moose/framework/src/materials/Material.C:132
#3  0x00007ffff52c6d9f in MaterialData::reinit<std::vector<std::shared_ptr<MaterialBase>, std::allocator<std::shared_ptr<MaterialBase> > > > (this=0x5555559353d0, mats=...)
    at /home/gburgio/projects/moose/framework/build/header_symlinks/MaterialData.h:454
#4  0x00007ffff526b677 in FEProblemBase::reinitMaterials (this=0x555555949050, blk_id=2, tid=0, swap_stateful=true)
    at /home/gburgio/projects/moose/framework/src/problems/FEProblemBase.C:3430
#5  0x00007ffff5df8df9 in NonlinearThread::onElement (this=0x7fffffffa5a0, elem=0x555555888ed0) at /home/gburgio/projects/moose/framework/src/loops/NonlinearThread.C:107
#6  0x00007ffff52d70f5 in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator() (this=0x7fffffffa5a0, 
    range=..., bypass_threading=false) at /home/gburgio/projects/moose/framework/build/header_symlinks/ThreadedElementLoopBase.h:226
#7  0x00007ffff63cd30b in libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeResidualThread> (
    range=..., body=...) at /home/gburgio/mambaforge3/envs/moose/libmesh/include/libmesh/threads_pthread.h:380
#8  0x00007ffff63a0592 in NonlinearSystemBase::computeResidualInternal (this=0x555555969e10, tags=...)
    at /home/gburgio/projects/moose/framework/src/systems/NonlinearSystemBase.C:1567
#9  0x00007ffff639a78f in NonlinearSystemBase::computeResidualTags (this=0x555555969e10, tags=...)
    at /home/gburgio/projects/moose/framework/src/systems/NonlinearSystemBase.C:789
#10 0x00007ffff5286ac6 in FEProblemBase::computeResidualTags (this=0x555555949050, tags=...) at /home/gburgio/projects/moose/framework/src/problems/FEProblemBase.C:6174
#11 0x00007ffff5285966 in FEProblemBase::computeResidualInternal (this=0x555555949050, soln=..., residual=..., tags=...)
    at /home/gburgio/projects/moose/framework/src/problems/FEProblemBase.C:6030
#12 0x00007ffff52842fc in FEProblemBase::computeResidual (this=0x555555949050, soln=..., residual=..., nl_sys_num=0)
    at /home/gburgio/projects/moose/framework/src/problems/FEProblemBase.C:5806
#13 0x00007ffff5284001 in FEProblemBase::computeResidualSys (this=0x555555949050, sys=..., soln=..., residual=...)
    at /home/gburgio/projects/moose/framework/src/problems/FEProblemBase.C:5771
#14 0x00007ffff6393431 in NonlinearSystem::solve (this=0x555555969e10) at /home/gburgio/projects/moose/framework/src/systems/NonlinearSystem.C:145
#15 0x00007ffff527fce6 in FEProblemBase::solve (this=0x555555949050, nl_sys_num=0) at /home/gburgio/projects/moose/framework/src/problems/FEProblemBase.C:5489
#16 0x00007ffff5db5f68 in FEProblemSolve::solve (this=0x5555559abbe8) at /home/gburgio/projects/moose/framework/src/executioners/FEProblemSolve.C:281
#17 0x00007ffff5dba910 in FixedPointSolve::solveStep (this=0x5555555f86b0, begin_norm=@0x555555abf5b0: 0, end_norm=@0x555555ab8740: 0, transformed_dofs=...)
    at /home/gburgio/projects/moose/framework/src/executioners/FixedPointSolve.C:426
#18 0x00007ffff5db98fa in FixedPointSolve::solve (this=0x5555555f86b0) at /home/gburgio/projects/moose/framework/src/executioners/FixedPointSolve.C:278
#19 0x00007ffff5dc1d0b in Steady::execute (this=0x5555559ab8c0) at /home/gburgio/projects/moose/framework/src/executioners/Steady.C:82
#20 0x00007ffff692f7e8 in MooseApp::executeExecutioner (this=0x5555556e3700) at /home/gburgio/projects/moose/framework/src/base/MooseApp.C:1113
#21 0x00007ffff693283e in MooseApp::run (this=0x5555556e3700) at /home/gburgio/projects/moose/framework/src/base/MooseApp.C:1418
#22 0x00005555555576a9 in main (argc=3, argv=0x7fffffffc288) at /home/gburgio/projects/perty/src/main.C:33
`

What is the cause of this? Do I need to somehow declare the number of equations as a global parameter in the input file to then initialize the vector parameter?

**Comment by gabburgio** [comment]
(I have also tried with 
 ` std::vector<Real> _vettore;`
in the header instead of 
`const std::vector<Real> & _vettore;`
to no avail)

**Comment by GiudGiud** [comment, marked as answer]
Hello

This is fine:
```
const std::vector<Real> & _vettore;
```
What's erroring is this
```
    _diff_vector[_qp][i] = _vettore[i];
```
because there's nowhere in the code where you resize this vector so that you can insert at this position.
You should add something like:
```
_diff_vector[_qp].resize(_vettore.size());
```
before the for loop.

One note:
A "vector" in MOOSE, such as a vector material or a vector variable, is a LIBMESH_DIM (usually 3) sized vector. 
A general sized vector is called an array instead. So you should look at array materials for example. 

Guillaume

**Reply by gabburgio** [reply]
Perfect, thanks for the answer and the pointers
